File: BusinessSDK/Sources/ToggleIdentitySDK/ToggleIdentitySDK.swift
import Foundation
import BusinessSDKCore

public enum ToggleIdentitySDKConfig {
    public static let baseURL = URL(string: "http://localhost:9011")!
}

public struct SendMagicLinkRequest: Codable, Sendable {
    public let email: String
    public let purpose: String

    public init(email: String, purpose: String) {
        self.email = email
        self.purpose = purpose
    }
}

public struct SendMagicLinkResponse: Codable, Sendable {
    public let message: String
    public let email: String
    public let purpose: String
    public let expiresIn: Int
    public let notificationStatus: String

    public init(message: String, email: String, purpose: String, expiresIn: Int, notificationStatus: String) {
        self.message = message
        self.email = email
        self.purpose = purpose
        self.expiresIn = expiresIn
        self.notificationStatus = notificationStatus
    }
}

public struct ApiResponseSendMagicLink: Codable, Sendable {
    public let success: Bool
    public let status: Int
    public let message: String
    public let data: SendMagicLinkResponse
    public let error: ErrorDetails
    public let timestamp: String
    public let traceId: String
    public let version: String

    public init(success: Bool, status: Int, message: String, data: SendMagicLinkResponse, error: ErrorDetails, timestamp: String, traceId: String, version: String) {
        self.success = success
        self.status = status
        self.message = message
        self.data = data
        self.error = error
        self.timestamp = timestamp
        self.traceId = traceId
        self.version = version
    }
}

public struct VerifyMagicLinkRequest: Codable, Sendable {
    public let email: String
    public let oobCode: String
    public let purpose: String
    public let deviceInfo: DeviceInfo

    public init(email: String, oobCode: String, purpose: String, deviceInfo: DeviceInfo) {
        self.email = email
        self.oobCode = oobCode
        self.purpose = purpose
        self.deviceInfo = deviceInfo
    }
}

public struct DeviceInfo: Codable, Sendable {
    public let deviceId: String
    public let platform: String
    public let appVersion: String

    public init(deviceId: String, platform: String, appVersion: String) {
        self.deviceId = deviceId
        self.platform = platform
        self.appVersion = appVersion
    }
}

public struct VerifyMagicLinkResponse: Codable, Sendable {
    public let idToken: String
    public let refreshToken: String
    public let tokenType: String
    public let expiresIn: Int
    public let user: UserInfo

    public init(idToken: String, refreshToken: String, tokenType: String, expiresIn: Int, user: UserInfo) {
        self.idToken = idToken
        self.refreshToken = refreshToken
        self.tokenType = tokenType
        self.expiresIn = expiresIn
        self.user = user
    }
}

public struct UserInfo: Codable, Sendable {
    public let userId: String
    public let email: String
    public let status: String

    public init(userId: String, email: String, status: String) {
        self.userId = userId
        self.email = email
        self.status = status
    }
}

public struct ApiResponseVerifyMagicLink: Codable, Sendable {
    public let success: Bool
    public let status: Int
    public let message: String
    public let data: VerifyMagicLinkResponse
    public let error: ErrorDetails
    public let timestamp: String
    public let traceId: String
    public let version: String

    public init(success: Bool, status: Int, message: String, data: VerifyMagicLinkResponse, error: ErrorDetails, timestamp: String, traceId: String, version: String) {
        self.success = success
        self.status = status
        self.message = message
        self.data = data
        self.error = error
        self.timestamp = timestamp
        self.traceId = traceId
        self.version = version
    }
}

public struct ErrorDetails: Codable, Sendable {
    public let code: String
    public let details: String
    public let fieldErrors: [FieldError]
    public let helpUrl: String

    public init(code: String, details: String, fieldErrors: [FieldError], helpUrl: String) {
        self.code = code
        self.details = details
        self.fieldErrors = fieldErrors
        self.helpUrl = helpUrl
    }
}

public struct FieldError: Codable, Sendable {
    public let field: String
    public let rejectedValue: String
    public let message: String

    public init(field: String, rejectedValue: String, message: String) {
        self.field = field
        self.rejectedValue = rejectedValue
        self.message = message
    }
}


private struct IdentityRepositorySendMagicLinkRequest: APIRequest {


    init() {

    }

    let method: HTTPMethod = .post
    let body: Data? = nil
    let headers: [String: String] = ["Accept": "application/json", "Content-Type": "application/json"]
    let requiresAuthorization: Bool = false

    var path: String {
        "/api/v1/identity/send-magic-link"
    }

    var queryItems: [URLQueryItem] {
        []
    }

    typealias Response = ApiResponseSendMagicLink
}

private struct IdentityRepositoryVerifyMagicLinkRequest: APIRequest {


    init() {

    }

    let method: HTTPMethod = .post
    let body: Data? = nil
    let headers: [String: String] = ["Accept": "application/json", "Content-Type": "application/json"]
    let requiresAuthorization: Bool = false

    var path: String {
        "/api/v1/identity/verify-magic-link"
    }

    var queryItems: [URLQueryItem] {
        []
    }

    typealias Response = ApiResponseVerifyMagicLink
}


public protocol IdentityRepositoryProtocol: Sendable {
    func sendMagicLink() async throws -> ApiResponseSendMagicLink
    func verifyMagicLink() async throws -> ApiResponseVerifyMagicLink
}

public final class IdentityRepository: IdentityRepositoryProtocol, @unchecked Sendable {
    private let networkEngine: NetworkEngineProtocol

    public init(networkEngine: NetworkEngineProtocol) {
        self.networkEngine = networkEngine
    }

public func sendMagicLink() async throws -> ApiResponseSendMagicLink {
    try await networkEngine.request(IdentityRepositorySendMagicLinkRequest())
}

public func verifyMagicLink() async throws -> ApiResponseVerifyMagicLink {
    try await networkEngine.request(IdentityRepositoryVerifyMagicLinkRequest())
}

}


public protocol ToggleIdentitySDKClientProtocol: Sendable {
    var identityRepository: IdentityRepositoryProtocol { get }
}

public struct ToggleIdentitySDKClient: ToggleIdentitySDKClientProtocol, Sendable {
    public let identityRepository: IdentityRepositoryProtocol

    public init(networkEngine: NetworkEngineProtocol = NetworkEngine(baseURL: ToggleIdentitySDKConfig.baseURL)) {
        self.identityRepository = IdentityRepository(networkEngine: networkEngine)
    }
}

File: android/modules/toggle-identity-sdk/src/main/java/com/toggle/identity/ToggleIdentitySDK.kt
package com.toggle.identity

import com.multiplatformbusinesssdk.core.ApiRequest
import com.multiplatformbusinesssdk.core.HttpMethod
import com.multiplatformbusinesssdk.core.NetworkEngine

object ToggleIdentitySDKConfig {
    const val baseUrl: String = "http://localhost:9011"
}

data class SendMagicLinkRequest(
    val email: String,
    val purpose: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): SendMagicLinkRequest {
                        val email = json.getString("email")
            val purpose = json.getString("purpose")
            return SendMagicLinkRequest(email, purpose)
        }
    }
}

data class SendMagicLinkResponse(
    val message: String,
    val email: String,
    val purpose: String,
    val expiresIn: Int,
    val notificationStatus: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): SendMagicLinkResponse {
                        val message = json.getString("message")
            val email = json.getString("email")
            val purpose = json.getString("purpose")
            val expiresIn = json.getInt("expiresIn")
            val notificationStatus = json.getString("notificationStatus")
            return SendMagicLinkResponse(message, email, purpose, expiresIn, notificationStatus)
        }
    }
}

data class ApiResponseSendMagicLink(
    val success: Boolean,
    val status: Int,
    val message: String,
    val data: SendMagicLinkResponse,
    val error: ErrorDetails,
    val timestamp: String,
    val traceId: String,
    val version: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): ApiResponseSendMagicLink {
                        val success = json.getBoolean("success")
            val status = json.getInt("status")
            val message = json.getString("message")
            val data = SendMagicLinkResponse.fromJson(json.getJSONObject("data"))
            val error = ErrorDetails.fromJson(json.getJSONObject("error"))
            val timestamp = json.getString("timestamp")
            val traceId = json.getString("traceId")
            val version = json.getString("version")
            return ApiResponseSendMagicLink(success, status, message, data, error, timestamp, traceId, version)
        }
    }
}

data class VerifyMagicLinkRequest(
    val email: String,
    val oobCode: String,
    val purpose: String,
    val deviceInfo: DeviceInfo
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): VerifyMagicLinkRequest {
                        val email = json.getString("email")
            val oobCode = json.getString("oobCode")
            val purpose = json.getString("purpose")
            val deviceInfo = DeviceInfo.fromJson(json.getJSONObject("deviceInfo"))
            return VerifyMagicLinkRequest(email, oobCode, purpose, deviceInfo)
        }
    }
}

data class DeviceInfo(
    val deviceId: String,
    val platform: String,
    val appVersion: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): DeviceInfo {
                        val deviceId = json.getString("deviceId")
            val platform = json.getString("platform")
            val appVersion = json.getString("appVersion")
            return DeviceInfo(deviceId, platform, appVersion)
        }
    }
}

data class VerifyMagicLinkResponse(
    val idToken: String,
    val refreshToken: String,
    val tokenType: String,
    val expiresIn: Int,
    val user: UserInfo
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): VerifyMagicLinkResponse {
                        val idToken = json.getString("idToken")
            val refreshToken = json.getString("refreshToken")
            val tokenType = json.getString("tokenType")
            val expiresIn = json.getInt("expiresIn")
            val user = UserInfo.fromJson(json.getJSONObject("user"))
            return VerifyMagicLinkResponse(idToken, refreshToken, tokenType, expiresIn, user)
        }
    }
}

data class UserInfo(
    val userId: String,
    val email: String,
    val status: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): UserInfo {
                        val userId = json.getString("userId")
            val email = json.getString("email")
            val status = json.getString("status")
            return UserInfo(userId, email, status)
        }
    }
}

data class ApiResponseVerifyMagicLink(
    val success: Boolean,
    val status: Int,
    val message: String,
    val data: VerifyMagicLinkResponse,
    val error: ErrorDetails,
    val timestamp: String,
    val traceId: String,
    val version: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): ApiResponseVerifyMagicLink {
                        val success = json.getBoolean("success")
            val status = json.getInt("status")
            val message = json.getString("message")
            val data = VerifyMagicLinkResponse.fromJson(json.getJSONObject("data"))
            val error = ErrorDetails.fromJson(json.getJSONObject("error"))
            val timestamp = json.getString("timestamp")
            val traceId = json.getString("traceId")
            val version = json.getString("version")
            return ApiResponseVerifyMagicLink(success, status, message, data, error, timestamp, traceId, version)
        }
    }
}

data class ErrorDetails(
    val code: String,
    val details: String,
    val fieldErrors: List<FieldError>,
    val helpUrl: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): ErrorDetails {
                        val code = json.getString("code")
            val details = json.getString("details")
            val fieldErrorsArray = json.getJSONArray("fieldErrors")
                        val fieldErrors = buildList {
                            for (index in 0 until fieldErrorsArray.length()) {
                                add(FieldError.fromJson(fieldErrorsArray.getJSONObject(index)))
                            }
                        }
            val helpUrl = json.getString("helpUrl")
            return ErrorDetails(code, details, fieldErrors, helpUrl)
        }
    }
}

data class FieldError(
    val field: String,
    val rejectedValue: String,
    val message: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): FieldError {
                        val field = json.getString("field")
            val rejectedValue = json.getString("rejectedValue")
            val message = json.getString("message")
            return FieldError(field, rejectedValue, message)
        }
    }
}


private class IdentityRepositorySendMagicLinkRequest() : ApiRequest<ApiResponseSendMagicLink> {
    override val path: String = "/api/v1/identity/send-magic-link"
    override val method: HttpMethod = HttpMethod.POST
    override val body: ByteArray? = null
    override val headers: Map<String, String> = mapOf("Accept" to "application/json", "Content-Type" to "application/json")
    override val query: Map<String, String> = emptyMap()
    override val requiresAuthorization: Boolean = false

    override fun parse(payload: String): ApiResponseSendMagicLink =
        ApiResponseSendMagicLink.fromJson(org.json.JSONObject(payload) as org.json.JSONObject)
}

private class IdentityRepositoryVerifyMagicLinkRequest() : ApiRequest<ApiResponseVerifyMagicLink> {
    override val path: String = "/api/v1/identity/verify-magic-link"
    override val method: HttpMethod = HttpMethod.POST
    override val body: ByteArray? = null
    override val headers: Map<String, String> = mapOf("Accept" to "application/json", "Content-Type" to "application/json")
    override val query: Map<String, String> = emptyMap()
    override val requiresAuthorization: Boolean = false

    override fun parse(payload: String): ApiResponseVerifyMagicLink =
        ApiResponseVerifyMagicLink.fromJson(org.json.JSONObject(payload) as org.json.JSONObject)
}


interface IdentityRepositoryProtocol {
    suspend fun sendMagicLink(): ApiResponseSendMagicLink
    suspend fun verifyMagicLink(): ApiResponseVerifyMagicLink
}

class IdentityRepository(
    private val networkEngine: NetworkEngine
) : IdentityRepositoryProtocol {
override suspend fun sendMagicLink(): ApiResponseSendMagicLink {
    return networkEngine.request(IdentityRepositorySendMagicLinkRequest())
}

override suspend fun verifyMagicLink(): ApiResponseVerifyMagicLink {
    return networkEngine.request(IdentityRepositoryVerifyMagicLinkRequest())
}

}


interface ToggleIdentitySDKClientProtocol {
    val identityRepository: IdentityRepositoryProtocol
}

class ToggleIdentitySDKClient(
    networkEngine: NetworkEngine = NetworkEngine(baseUrl = ToggleIdentitySDKConfig.baseUrl)
) : ToggleIdentitySDKClientProtocol {
    override val identityRepository: IdentityRepositoryProtocol = IdentityRepository(networkEngine)
}
